future-apps/agent/contracts/SignatureValidator.sol

```diff
@@ -42,18 +80,58 @@ library SignatureValidator {
         }
 
         if (v != 27 && v != 28) {
-            return false;
+            badSig = true;
         }
+    }
 
-        bytes32 signedHash;
-        if (mode == SignatureMode.EthSign) {
-            signedHash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hash));
-        } else if (mode == SignatureMode.EIP712) {
-            signedHash = hash;
-        } else {
-            return false;
+    /**
+     * @dev DANGEROUS FUNCTION: Modifies input in place, making it invalid after using this function
+     * Memory layout: bytes in parenthesis are returned
+     * L = l - 1
+     * Input:  ([l1][l2]....[l32][b1][b2]....[bn])
+     * Output: [00]([L1][L2]....[L32][b2]....[bn])
+     */
+    function popFirstByte(bytes memory input) internal pure returns (bytes memory output, byte firstByte) {
+        assembly {
+            let length := mload(input)
+            firstByte := mload(add(input, 0x20))
+            output := add(input, 0x01)
```

Comment:
This is pretty scary haha. Basically had to run this through remix to see what it would actually do to the memory.

I can see why people prefer to pop the last byte instead of the first byte now (perhaps we should move the mode to the end too?).

Reply:
Will modify to copying into a new array for safety

Reply:
Solved in 0d3e2ecde4bd86f90dca8959f2e36efa186d4bc2